### PR TITLE
feat: add installation chart framework

### PR DESCRIPTION
**Pull request by thxCode**
**Fixes:**

#10

**Problem:**

There is no easy approach to install Harvester.

**Solution:**

Give a Helm chart installation solution.

- `harvester` chart is used to install `kubevirt`, `cdi`, `minio`, and `harvester`.
- we can disable each component individually.
- The `minio` chart is a copy from the official chart.
  ``` bash
  $ cd deploy/charts/harvester
  $ helm repo add minio https://helm.min.io/
  $ helm pull minio/minio
  $ tar -xvzf minio-*.tgz -C charts
  $ rm minio-*.tgz
  ```
- The parameter of `kubevirt` and `cdi` charts have been sorted out, and kept consistent with the parameters of CRD as much as possible and described.

When `KubeVirt`/`CDI` resource is deleted, the admission webhook of `KubeVirt Operator`/`CDI Operator` workload will be triggered, and the related resources cannot be cleaned up if the corresponding Operator workload has been deleted. In order to avoid this, we use a `pre-delete` Helm hook to clean up the `KubeVirt`/`CDI` resource before executed `helm uninstall`.

- Installation

  ```bash
  $ # create namespace
  $ kubectl create namespace cattle-vm
  
  $ # install harvester with emulation if needed
  $ helm install harvester deploy/charts/harvester --namespace cattle-vm --set-string 
  kubevirt.spec.configuration.developerConfiguration.useEmulation=true
  ```

- Uninstallation

  ```bash
  $ # uninstall harvester
  $ helm uninstall harvester--namespace cattle-vm
  ```

**Test plan:**

- `helm install ...`

### CODE REVIEW

**Comment by thxCode** [comment]
Since `rancher/vm` is still working in progress, this PR is a basic installation chart framework, and the default rendering configuration of these sub-charts needs further PR.

File: deploy/charts/harvester/templates/job.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            - "/bin/bash"
+          args:
+            - "-c"
+{{- if and (index .Values "kubevirt-operator" "enabled") (index .Values "cdi-operator" "enabled") }}
+            - "echo '[INFO] deleting cdi ...'; kubectl delete cdi --wait --all; echo '[INFO] deleting kubevirt ...'; kubectl delete kubevirt --wait --all; exit"
+{{- else if index .Values "kubevirt-operator" "enabled" }}
+            - "echo '[INFO] deleting kubevirt ...'; kubectl delete kubevirt --wait --all; exit"
+{{- else }}
+            - "echo '[INFO] deleting cdi ...'; kubectl delete cdi --wait --all; exit"
+{{- end }}
```
Comment (gitlawr) [inline deploy/charts/harvester/templates/job.yaml]: Looks like two conditions are sufficient for this
```
if(kvo enabled) delete; endif; 
if(cdio enabled) delete; endif;
```

File: deploy/charts/harvester/values.yaml [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        cpu: 500m
+        memory: 500Mi
+      requests:
+        cpu: 250m
+        memory: 256Mi
+
+    ## Specify the gateway to access Minio-like VM image storage service which is compatible with S3,
+    ## only valid if Minio sub-charts is disabled in the same chart release.
+    ##
+    vmImageStorageGateway:
```
Comment (gitlawr) [inline deploy/charts/harvester/values.yaml:167]: Is it a common word using `gateway` in this use case?

File: deploy/charts/harvester/README.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+- Deploy a KubeVirt Operator if needed, defaults to deploy.
+- Deploy a KubeVirt CRD resource to enable KubeVirt if needed, defaults to deploy.
+- Deploy a KubeVirt Containerized Data Importer Operator if needed, defaults to deploy.
+- Deploy a CDI CRD resource to enable KubeVirt Containerized Data Importer if needed, default to deploy.
+- Deploy a Minio as virtual machine image storage if needed, defaults to deploy.
+- Deploy the Harvester resources.
+
+### Prerequisites
+
+- Kubernetes 1.16+ with Beta APIs enabled.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/README.md]: Outline which version of Kubernetes we can support.

File: deploy/charts/harvester/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+- Helm 3.2+.
+- KubeVirt Operator 0.30.5+.
+- KubeVirt Containerized Data Importer Operator 1.21.0+.
+- Image storage compatible with S3 like Minio 5.0.32+.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/README.md]: These components are the optional installation in this chart, I think it is better to indicate these, I will use a new section to outline this.

File: deploy/charts/harvester/README.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+### Installing the Chart
+
+To install the chart with the release name `harvester`.
+
+```bash
+$ # create target namespace
+$ kubectl create ns cattle-vm
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/README.md]: Or `harvester-system`

[the remark this answers is not in the document]
Comment (guangbochen) [reply deploy/charts/harvester/README.md]: this also works.

File: deploy/charts/harvester/values.yaml (hunk 2 above)
Comment (thxCode) [reply deploy/charts/harvester/values.yaml:167]: Inspired by Minio values.

File: deploy/charts/harvester/charts/kubevirt/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+apiVersion: v2
+name: kubevirt
+version: 0.1.0
+appVersion: v1alpha3
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/charts/kubevirt/Chart.yaml:4]: `KubeVirt` and `CDI` are only a chart for CRD resource, keep the `appVersion` as the same as the CRD version is an approach to show the version of this CRD resource.

File: deploy/charts/harvester/templates/job.yaml (hunk 1 above)
Comment (thxCode) [reply deploy/charts/harvester/templates/job.yaml]: Yes, in fact, they are only two conditions. However, here needs to write in one line behind `/bin/bash -c`. I have merged these conditions in one line.

File: deploy/charts/harvester/README.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+- Deploy a KubeVirt Operator if needed, defaults to deploy.
+- Deploy a KubeVirt CRD resource to enable KubeVirt if needed, defaults to deploy.
+- Deploy a KubeVirt Containerized Data Importer Operator if needed, defaults to deploy.
+- Deploy a CDI CRD resource to enable KubeVirt Containerized Data Importer if needed, default to deploy.
+- Deploy a Minio as virtual machine image storage if needed, defaults to deploy.
+- Deploy the Harvester resources.
+
+### Prerequisites
+
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/README.md:17]: Updated

File: deploy/charts/harvester/templates/NOTES.txt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+The Harvester has been installed into "{{ .Release.Namespace }}" namespace with "{{ .Release.Name }}" as release name.
+
+- {{ if index .Values "kubevirt-operator" "enabled" }}[x]{{ else }}[ ]{{ end }} KubeVirt Operator
+- {{ if .Values.kubevirt.enabled }}[x]{{ else }}[ ]{{ end }} KubeVirt Resource named "kubevirt"
+- {{ if index .Values "cdi-operator" "enabled" }}[x]{{ else }}[ ]{{ end }} KubeVirt Containerized Data Importer Operator
+- {{ if .Values.cdi.enabled }}[x]{{ else }}[ ]{{ end }} KubeVirt Containerized Data Importer(CDI) Resource named "cdi"
+- {{ if .Values.minio.enabled }}[x]{{ else }}[ ]{{ end }} Minio
+
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/templates/NOTES.txt:8]: Updated

File: deploy/charts/harvester/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,154 @@
+# NB(thxCode): Get the default fully qualified name of Minio,
+# the below logic is from "charts/minio/templates/_helpers.tpl",
+# and then we indicate the values from "minio" domain.
+{{- define "charts.minio.fullname" }}
+{{- if .Values.minio.fullnameOverride -}}
+{{- .Values.minio.fullnameOverride | trunc 63 | trimSuffix "-" -}}
+{{- else -}}
+{{- $name := default "minio" .Values.minio.nameOverride -}}
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/templates/deployment.yaml]: Yes, updated

File: deploy/charts/harvester/README.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+If you don't want to install KubeVirt Containerized Data Importer CRD resource, you can do the following.
+
+```bash
+$ helm install harvester harvester --namespace cattle-vm \
+    --set cdi.enabled=false
+```
+
+#### Configure Minio
+
+> **Noted:** We made a [Minio chart](https://github.com/minio/charts) backup under `charts` directory.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/README.md]: Updated

File: deploy/charts/harvester/README.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+$ # create target namespace
+$ kubectl create ns cattle-vm
+
+$ # install chart to target namespace
+$ helm install harvester harvester --namespace cattle-vm
+```
+
+### Uninstalling the Chart
+
+To uninstall/delete the `harvester` deployment.
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/README.md]: Updated

File: deploy/charts/harvester/templates/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+kind: Deployment
+metadata:
+  # NB(thxCode): name should not be customized as below:
+  # name: {{ template "harvester.fullname" . }}
+  # because we can easily confirm this resource from the corresponding namespace.
+  name: harvester
+  labels:
+{{ include "harvester.labels" . | indent 4 }}
+    app.kubernetes.io/name: apiserver
+    app.kubernetes.io/component: workload
```
[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/templates/deployment.yaml]: Updated

File: deploy/charts/harvester/charts/cdi-operator/templates/serviceaccount.yaml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  # NB(thxCode): name cannot be customized as below:
+  # name: {{ template "cdi-operator.fullname" . }}
+  # because the installation strategy dump action needs this ServiceAccount.
+  name: cdi-operator
+  labels:
+{{ include "cdi-operator.labels" . | indent 4 }}
+    app.kubernetes.io/component: metadata
```
Comment (guangbochen) [inline deploy/charts/harvester/charts/cdi-operator/templates/serviceaccount.yaml]: please resolve the newline, and prefer to use `rbac` of its component value, metadata sounds wired.

File: deploy/charts/harvester/charts/cdi/README.md [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+- Helm 3.2+.
+- KubeVirt Containerized Data Importer Operator v1.21.0+ with Beta APIs enabled.
+
+### Installing the Chart
+
+To install the chart with the release name `my-release`.
+
+```bash
+$ # install chart to target namespace installed cdi-operator, i.e: cdi-system
+$ helm install my-release cdi --namespace cdi-system
```
Comment (guangbochen) [inline deploy/charts/harvester/charts/cdi/README.md]: I don't know whether the user will want to install the CDI chart independently, but I would prefer to use `harvester-system` for better consistency.

File: deploy/charts/harvester/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (gitlawr) [reply deploy/charts/harvester/README.md]: I don't understand what `with Beta APIs enabled` refers to.  Let's only specify the kubernetes version

[the remark this answers is not in the document]
Comment (thxCode) [reply deploy/charts/harvester/README.md]: Updated

File: deploy/charts/harvester/charts/cdi/README.md (hunk 14 above)
Comment (thxCode) [reply deploy/charts/harvester/charts/cdi/README.md]: Updated

File: deploy/charts/harvester/charts/cdi-operator/templates/serviceaccount.yaml (hunk 13 above)
Comment (thxCode) [reply deploy/charts/harvester/charts/cdi-operator/templates/serviceaccount.yaml]: I don't think ServiceAccount is a thing for RBAC, it is used to connect RBAC, it is a metadata described what the service account is. ~I will change back to the previous name: permission, this seems more reasonable.~

**Comment by thxCode** [comment]
Projected all Kubernetes application labels to their related workload, so we don't need to use a `component` description to group/separate Role/Binding/ServiceAccount/..., they can connect to the corresponding workload with these labels. cc @guangbochen , @gitlawr

**Review by guangbochen — APPROVED** [review]
LGTM, thanks

**Review by gitlawr — APPROVED** [review]

**Pull request merged**